File: packages/agents/src/codex.ts
/**
 * Codex adapter.
 *
 * Path and hook choices are isolated to this module: when Codex's real on-disk
 * layout is confirmed, only the values below change, never their consumers.
 *
 * - Skills: project `<project>/.codex/skills/<name>/`, global
 *   `~/.codex/skills/<name>/`.
 * - Hooks: `json-merge` into `<base>/.codex/settings.json` under `hooks`.
 *   Codex stores configuration as a structured settings file with no comment
 *   syntax, so JSON merge with an ownership marker is the appropriate strategy.
 */

import type { AgentTarget, HostEnv } from '@skillkeeper/core';
import { makeAdapter } from './makeAdapter.js';
import { baseDir, joinPath } from './paths.js';

function codexDir(target: AgentTarget, env: HostEnv): string {
  return joinPath(baseDir(target, env), '.codex');
}

export const codexAdapter = makeAdapter({
  kind: 'codex',
  skillsRoot: (target, env) => joinPath(codexDir(target, env), 'skills'),
  availabilityDir: (env) => joinPath(env.homeDir, '.codex'),
  hook: {
    strategy: 'json-merge',
    async resolveTargetFile(target: AgentTarget, env: HostEnv): Promise<string> {
      return joinPath(codexDir(target, env), 'settings.json');
    },
  },
});

File: crates/skillkeeper-agents/src/codex.rs
//! Codex adapter. Ported from `packages/agents/src/codex.ts`.
//!
//! - Skills: project `<project>/.codex/skills/<name>/`, global
//!   `~/.codex/skills/<name>/`.
//! - Hooks: `json-merge` into `<base>/.codex/settings.json` under `hooks`.

use skillkeeper_core::ports::{HostEnv, PortResult};

use crate::adapter::{make_adapter, AdapterSpec, AgentAdapter, HookCapability};
use crate::model::{AgentKind, AgentTarget, HookStrategy};
use crate::paths::{base_dir, join_path};

fn codex_dir(target: &AgentTarget, env: &dyn HostEnv) -> PortResult<String> {
    Ok(join_path(&[&base_dir(target, env)?, ".codex"]))
}

/// Build the Codex adapter.
pub fn codex_adapter() -> AgentAdapter {
    make_adapter(AdapterSpec {
        kind: AgentKind::Codex,
        skills_root: Box::new(|target, env| Ok(join_path(&[&codex_dir(target, env)?, "skills"]))),
        availability_dir: Box::new(|env| join_path(&[env.home_dir(), ".codex"])),
        discovery_group: None,
        guidance_file: Box::new(|_fs, target, env| {
            Ok(join_path(&[&base_dir(target, env)?, "AGENTS.md"]))
        }),
        hook: HookCapability::new(
            HookStrategy::JsonMerge,
            None,
            None,
            Box::new(|target, env| Ok(join_path(&[&codex_dir(target, env)?, "settings.json"]))),
        ),
    })
}
